pageserver/src/tenant/layer_map.rs

```diff
@@ -332,28 +363,67 @@ where
             "expected and new must both be l0 deltas or neither should be: {expected_l0} != {new_l0}"
         );
 
-        let l0_index = if expected_l0 {
-            // find the index in case replace worked, we need to replace that as well
-            Some(
-                self.l0_delta_layers
-                    .iter()
-                    .position(|slot| Self::compare_arced_layers(slot, expected))
-                    .ok_or_else(|| anyhow::anyhow!("existing l0 delta layer was not found"))?,
-            )
+        let lr = expected.get_lsn_range();
+        let replaced = self.historic_layers.remove(&LayerRef(expected.clone()));
+        if replaced {
+            for kr in expected.get_occupied_ranges(ctx)? {
+                match self.historic.replace(
+                    &historic_layer_coverage::LayerKey {
+                        key: kr.start.to_i128()..kr.end.to_i128(),
+                        lsn: lr.start.0..lr.end.0,
+                        is_image: !expected.is_incremental(),
+                    },
+                    new.clone(),
+                    |existing| Self::compare_arced_layers(existing, expected),
+                ) {
+                    Replacement::Replaced { .. } => { /* expected */ }
+                    Replacement::NotFound => {
+                        // TODO: the downloaded file should probably be removed, otherwise
+                        // it will be added to the layermap on next load? we should
+                        // probably restart any get_reconstruct_data search as well.
+                        //
+                        // See: https://github.com/neondatabase/neon/issues/3533
+                        anyhow::bail!("replacing downloaded layer into layermap failed because layer was not found");
+                    }
+                    Replacement::RemovalBuffered => {
+                        unreachable!("current implementation does not remove anything")
+                    }
+                    Replacement::Unexpected(_other) => {
+                        // if the other layer would have the same pointer value as
+                        // expected, it means they differ only on vtables.
+                        //
+                        // otherwise there's no known reason for this to happen as
+                        // compacted layers should have different covering rectangle
+                        // leading to produce Replacement::NotFound.
+
+                        anyhow::bail!(
+                            "replacing downloaded layer into layermap failed because another layer was found instead of expected"
+                        );
+                    }
```

Comment:
I am quite unsure of the outcome here (making these all errors) at least for on-demand download. I will investigate this more in later PR, so this will have to adapt to those changes.